tsfc/coffee_mode.py

```diff
@@ -0,0 +1,282 @@
+    big = 1e20  # some arbitrary big number
+    ilp_prob += ilp.lpSum(ilp_var[index] * (big - index_extent(atomic, argument_indices))
+                          for atomic, index in iteritems(atomic_index))
+
+    # constraints
+    for connection in connections:
+        ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
+
+    ilp_prob.solve()
+    if ilp_prob.status != 1:
+        raise RuntimeError("Something bad happened during ILP")
+
+    def optimal(atomic):
+        return ilp_var[atomic_index[atomic]].value() == 1
+
+    optimal_atomics = filter(optimal, iterkeys(atomic_index))
+    return tuple(optimal_atomics)
+
+
+def factorise_atomics(monomials, optimal_atomics, argument_indices):
+    """Group and factorise monomials using a list of atomics as common
+    subexpressions. Create new monomials for each group and optimise them recursively.
+
+    :arg monomials: an iterable of :class:`Monomial`s, all of which should have
+                    the same sum indices
+    :arg optimal_atomics: list of tuples of atomics to be used as common subexpression
+    :arg argument_indices: tuple of argument indices
+
+    :returns: an iterable of factorised :class:`Monomials`s, or the original
+              input if no changes are made.
+    """
+    if not optimal_atomics or len(monomials) < 2:
+        return monomials
+
+    # Group monomials with respect to each optimal atomic
+    def group_key(monomial):
+        for oa in optimal_atomics:
+            if oa in monomial.atomics:
+                return oa
+        assert False, "Expect at least one optimal atomic per monomial."
+    factor_group = groupby(monomials, key=group_key)
+
+    # We should not drop monomials
+    assert sum(len(ms) for _, ms in factor_group) == len(monomials)
+
+    sum_indices = next(iter(monomials)).sum_indices
+    new_monomials = []
+    for oa, monomials in factor_group:
+        # Create new MonomialSum for the factorised out terms
+        sub_monomials = []
+        for monomial in monomials:
+            atomics = list(monomial.atomics)
+            atomics.remove(oa)  # remove common factor
+            sub_monomials.append(Monomial((), tuple(atomics), monomial.rest))
+        sub_monomials = optimise_monomials(sub_monomials, argument_indices)
+        assert len(sub_monomials) > 0
+        if len(sub_monomials) == 1:
+            # result is a product, add back the common atomics then add to
+            # new MonomialSum directly
+            sub_monomial, = sub_monomials
+            new_monomials.append(
+                Monomial(sum_indices, sub_monomial.atomics + (oa,), sub_monomial.rest))
+        else:
+            # result is a sum, we need to create a new node
+            node = monomial_sum_to_expression(sub_monomials)
+            if set(argument_indices) & set(node.free_indices):
+                new_monomials.append(Monomial(sum_indices, (oa, node), one))
```

Comment:
The terminology `collect_monomials` uses with @tj-sun's `classifier`: if node's free indices have intersection with the argument indices, then it's either ATOMIC (only one argument index) or COMPOUND (more than one argument index), otherwise it's OTHER.